src/Passes/Source/Rules/Factory.cpp

```diff
@@ -502,6 +507,82 @@ namespace quantum
              replace_branch_positive});
     }
 
+    void RuleFactory::optimiseSelectQuantumOne()
+    {
+        auto replace_select_positive = [](Builder& builder, Value* val, Captures& cap, Replacements& replacements) {
+            auto result = cap["result"];
+            auto cond   = llvm::dyn_cast<llvm::Instruction>(cap["cond"]);
+            // Replacing result
+            auto orig_instr = llvm::dyn_cast<llvm::Instruction>(val);
+            if (orig_instr == nullptr)
+            {
+                return false;
+            }
+
+            auto                      module   = orig_instr->getModule();
+            auto                      function = module->getFunction("__quantum__qir__read_result");
+            std::vector<llvm::Value*> arguments;
+            arguments.push_back(result);
+
+            if (!function)
+            {
+                std::vector<llvm::Type*> types;
+                types.resize(arguments.size());
+                for (uint64_t i = 0; i < types.size(); ++i)
+                {
+                    types[i] = arguments[i]->getType();
+                }
+
+                auto return_type = llvm::Type::getInt1Ty(val->getContext());
+
+                llvm::FunctionType* fnc_type = llvm::FunctionType::get(return_type, types, false);
+                function                     = llvm::Function::Create(
+                    fnc_type, llvm::Function::ExternalLinkage, "__quantum__qir__read_result", module);
+            }
+
+            builder.SetInsertPoint(llvm::dyn_cast<llvm::Instruction>(val));
+            auto new_call = builder.CreateCall(function, arguments);
+            if (cond == nullptr)
+            {
+                return false;
+            }
+
+            new_call->takeName(cond);
+
+            for (auto& use : cond->uses())
+            {
+                llvm::User* user = use.getUser();
+                user->setOperand(use.getOperandNo(), new_call);
+            }
+            cond->replaceAllUsesWith(new_call);
+
+            // Deleting the previous condition and function to fetch one
+            replacements.push_back({cond, nullptr});
+            replacements.push_back({cap["one"], nullptr});
+
+            return false;
+        };
```

Comment:
In all the cases the lambda returns `false`?

Reply:
That is a good point. The function should return `true` here to indicate that the replacement was successful.